ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java

```diff
@@ -95,67 +96,70 @@ public Gas gasRequirement(final BytesValue input) {
-
-      WorldUpdater publicWorldState = messageFrame.getWorldState();
-
-      BytesValue privacyGroupId =
-          BytesValue.wrap(receiveResponse.getPrivacyGroupId().getBytes(Charsets.UTF_8));
-      // get the last world state root hash - or create a new one
-      Hash lastRootHash =
-          privateStateStorage.getPrivateAccountState(privacyGroupId).orElse(EMPTY_ROOT_HASH);
-      MutableWorldState disposablePrivateState =
-          privateWorldStateArchive.getMutable(lastRootHash).get();
-
-      WorldUpdater privateWorldStateUpdater = disposablePrivateState.updater();
-
-      TransactionProcessor.Result result =
-          privateTransactionProcessor.processTransaction(
-              messageFrame.getBlockchain(),
-              publicWorldState,
-              privateWorldStateUpdater,
-              messageFrame.getBlockHeader(),
-              privateTransaction,
-              messageFrame.getMiningBeneficiary(),
-              OperationTracer.NO_TRACING,
-              messageFrame.getBlockHashLookup(),
-              privacyGroupId);
-
-      if (result.isInvalid() || !result.isSuccessful()) {
-        throw new Exception("Unable to process the private transaction");
-      }
+      receiveResponse = enclave.receive(receiveRequest);
+    } catch (IOException e) {
+      LOG.debug("Enclave probably does not have private transaction with key {}.", key);
+      return BytesValue.EMPTY;
+    }
+
+    final BytesValueRLPInput bytesValueRLPInput =
+        new BytesValueRLPInput(
+            BytesValue.wrap(Base64.getDecoder().decode(receiveResponse.getPayload())), false);
+
+    final PrivateTransaction privateTransaction = PrivateTransaction.readFrom(bytesValueRLPInput);
+
+    final WorldUpdater publicWorldState = messageFrame.getWorldState();
+
+    final BytesValue privacyGroupId =
+        BytesValue.wrap(receiveResponse.getPrivacyGroupId().getBytes(Charsets.UTF_8));
+    // get the last world state root hash - or create a new one
+    final Hash lastRootHash =
+        privateStateStorage.getPrivateAccountState(privacyGroupId).orElse(EMPTY_ROOT_HASH);
+
+    final MutableWorldState disposablePrivateState =
+        privateWorldStateArchive.getMutable(lastRootHash).get();
+
+    final WorldUpdater privateWorldStateUpdater = disposablePrivateState.updater();
+    final PrivateTransactionProcessor.Result result =
+        privateTransactionProcessor.processTransaction(
+            messageFrame.getBlockchain(),
+            publicWorldState,
+            privateWorldStateUpdater,
+            messageFrame.getBlockHeader(),
+            privateTransaction,
+            messageFrame.getMiningBeneficiary(),
+            new DebugOperationTracer(TraceOptions.DEFAULT),
+            messageFrame.getBlockHashLookup(),
+            privacyGroupId);
+
+    if (result.isInvalid() || !result.isSuccessful()) {
+      LOG.error("Unable to process the private transaction: {}", result.getValidationResult());
```

Comment:
Should we log more context about what went wrong? Is the validationResult enough to work with if we need to debug this?

Reply:
yes the `validationResult` is the reason the `result` is "invalid" or "unsuccessful"